### DISCUSSION TITLE
How to detect if a workflow node is stuck in pending state for a prolonged time?

### DISCUSSION THREAD

**Discussion by lpaladin** [argoproj/argo-workflows#11746] [Q&A] [answered]
Hello! I'd like to know if there is any existing functionality to detect if a pod created in workflow is stuck in pending state? I asked because I want to distinguish between stuck in "ImagePullBackOff" and running normally.

I tried hooks to get notifications when node state changes, but since hooks is only called upon changes, I assume I'm unable to achieve this goal?

Thank you in advance!

**Comment by agilgur5** [comment, marked as answer]
You seem to be on the right track with most of your questions! I haven't worked with hooks too much, but as far as I know, I think you'd be correct that you can't directly detect that something is _stuck_ in `Pending`. You can detect that it hasn't moved out of `Pending` (i.e. the inverse).

Using a timeout via [`activeDeadlineSeconds`](https://argoproj.github.io/argo-workflows/walk-through/timeouts/) is typically a way around that.
Might be able to detect it with [metrics](https://argoproj.github.io/argo-workflows/metrics/) as well.

**Reply by lpaladin** [reply]
Thank you for pointing out [`activeDeadlineSeconds`](https://argoproj.github.io/argo-workflows/walk-through/timeouts/) and [metrics](https://argoproj.github.io/argo-workflows/metrics/)! Seems metrics can do the work with Real-Time Metrics + labels. 👍🏻

A quick follow-up question: Can I specify `activeDeadlineSeconds` for `Pending` phase only? It would be ideal to only abort in `Pending` but not `Running`...

**Reply by agilgur5** [reply]
Interesting question. That is not currently possible. It can only be set Workflow or Template-level and has no condition.

I was gonna say that conditional `activeDeadlineSeconds` might be a useful feature, but that could create some wacky edge cases that don't exist when it is static (for instance, if a status changes and now it has _already_ passed its deadline)

**Reply by lpaladin** [reply]
Thank you for clarifying!
